noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,309 @@
+        let shield_owner_npk_m_hash = get_current_public_keys(&mut context, shield_owner).npk_m.hash();
+        let shield_note_slot = storage.private_nfts.at(shield_owner).storage_slot;
+        let hiding_point = NFTNoteHidingPoint::new(shield_owner_npk_m_hash, shield_note_slot, note_randomness);
+
+        // We make the owner part of the slot preimage to ensure he cannot interfere with non-owner's slots
+        let owner_storage_slot_commitment: Field = pedersen_hash(
+            [shield_owner.to_field(), transient_storage_slot_randomness],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+        // Then we hash the owner storage slot commitment with finalizer and use that as the final shield slot
+        // --> by hashing it with a finalizer we ensure that finalizer cannot interfere with slots not assigned to him.
+        let slot: Field = pedersen_hash(
+            [finalizer.to_field(), owner_storage_slot_commitment],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+
+        NFT::at(context.this_address()).store_shield_in_transient_storage(hiding_point, slot).enqueue(&mut context);
+    }
+
+    #[aztec(public)]
+    #[aztec(internal)]
+    fn store_shield_in_transient_storage(point: NFTNoteHidingPoint, slot: Field) {
+        // We don't perform check for the overwritten value to be non-zero because the slots are siloed to the shield
+        // owner and hence the owner can interfere only with his own execution.
+        context.storage_write(slot, point);
+    }
+
+    #[aztec(public)]
+    fn send_to_shield(
+        from: AztecAddress,
+        token_id: Field,
+        owner_storage_slot_commitment: Field,
+        nonce: Field
+    ) {
+        // Perform authwit, owner and prepared shield checks
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let current_owner = storage.public_owners.at(token_id).read();
+        assert(current_owner.eq(from), "invalid NFT owner");
+
+        // Derive the shield slot from the owner storage slot commitment and the finalizer address (declared
+        // as `from` in this function)
+        let shield_slot = pedersen_hash(
+            [from.to_field(), owner_storage_slot_commitment],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+
+        // Read the hiding point from "transient" storage and check it's not empty to ensure the shield was prepared
+        let mut hiding_point: NFTNoteHidingPoint = context.storage_read(shield_slot);
+        assert(!is_empty(hiding_point), "shield not prepared");
+
+        // Set the original NFT owner to zero
+        storage.public_owners.at(token_id).write(AztecAddress::zero());
+
+        // Read the NFT hiding point from "transient" storage and add the token ID to it
+        hiding_point.add_token_id(token_id);
```

Comment:
Isn't this unnecessary? We could have computed the entire note in private and just sent the note hash, since we know all information. Things are diffrent in e.g. the FPC because in that case we don't know the amount during private execution, but here we do.

Reply:
You might be right. TBH I was trying to demonstrate the proposed transient storage shield flow and didn't think through whether it's actually needed for the NFT. We for sure need that for the token because then it allows for the uniswap flow (prepare shield, swap tokens in public call, only after the swap the token amount is known, finalize shield).

Don't you think this might be useful even for the NFT as it essentially gives some external contract control over what NFT we will receive? The only app that currently comes to my mind is some kind of lottery when you randomly swap your NFT (although not sure if we'll even have randomness available in public, aztec randao?)

@rahul-kothari you seem to have spent quite some time thinking about apps. Shall we keep the partial notes transient storage shield flow here or is it useless?

@nventuro would you approve of this flow for the token contract? If yes I would go ahead and refactor token contract to use it. As far as I know it's now clear that we'll go the partial notes transient storage way.

Reply:
Even if we don't have the flow, we still need to use transient storage to pass the completed note, don't we? We don't need partial notes, but we'd still be dealing with the contract having to communicate with tiself.

I'd keep this as is for now and we can then evaluate how we add it to the fungible token case.